File: android/src/main/java/expo/modules/sonycamera/SonyCameraModule.kt
package expo.modules.sonycamera

import android.content.Intent
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class SonyCameraModule : Module(), SonyCameraController.Listener {
  private lateinit var controller: SonyCameraController

  override fun definition() = ModuleDefinition {
    Name("SonyCamera")

    Events("onStateChanged", "onDeviceAttached", "onPhotoCaptured")

    OnCreate {
      val context = requireNotNull(appContext.reactContext)
      controller = SonyCameraController(context.applicationContext)
      controller.addListener(this@SonyCameraModule)
      SonyCameraRegistry.controller = controller
      appContext.currentActivity?.intent?.let(controller::handleIntent)
      controller.refreshAttachedDevice(autoConnect = true)
    }

    OnNewIntent { intent -> controller.handleIntent(intent) }
    OnActivityEntersForeground { controller.refreshAttachedDevice(autoConnect = true) }

    OnDestroy {
      controller.removeListener(this@SonyCameraModule)
      controller.close()
      if (SonyCameraRegistry.controller === controller) SonyCameraRegistry.controller = null
    }

    Function("getState") { controller.statePayload() }

    // The options argument is declared in TypeScript, so it has to be accepted here even
    // though candidate selection is not implemented yet. Expo validates argument count,
    // so a zero-argument definition made the documented signature throw at runtime.
    AsyncFunction("connect") { options: Map<String, Any?>? ->
      controller.connectBlocking(options.orEmpty())
    }
    AsyncFunction("disconnect") { controller.disconnectBlocking() }
    Function("getDiagnostics") { controller.diagnosticsSnapshot() }
    Function("clearDiagnostics") { controller.clearDiagnostics() }
    AsyncFunction("startLiveView") { controller.startLiveView(); controller.statePayload() }
    AsyncFunction("stopLiveView") { controller.stopLiveView(); controller.statePayload() }
    AsyncFunction("capturePhoto") { controller.capturePhotoBlocking() }
    AsyncFunction("capturePreviewFrame") { controller.capturePreviewFrameBlocking() }
    AsyncFunction("focusAt") { x: Double, y: Double, viewWidth: Double, viewHeight: Double ->
      controller.focusAtBlocking(x, y, viewWidth, viewHeight)
    }

    View(SonyCameraView::class) {
      Prop("active") { view: SonyCameraView, active: Boolean -> view.setActive(active) }
      OnViewDestroys { view: SonyCameraView -> view.cleanup() }
    }
  }

  override fun onStateChanged(payload: Map<String, Any?>) {
    sendEvent("onStateChanged", payload)
  }

  override fun onDeviceAttached(payload: Map<String, Any?>) {
    sendEvent("onDeviceAttached", payload)
  }

  override fun onPhotoCaptured(payload: Map<String, Any?>) {
    sendEvent("onPhotoCaptured", payload)
  }

  override fun onFrame(jpeg: ByteArray) = Unit
}

internal object SonyCameraRegistry {
  @Volatile var controller: SonyCameraController? = null
}

File: ios/SonyCameraModule.swift
import ExpoModulesCore

public final class SonyCameraModule: Module, SonyCameraControllerListener {
  private var controller: SonyCameraController!

  public func definition() -> ModuleDefinition {
    Name("SonyCamera")

    Events("onStateChanged", "onDeviceAttached", "onPhotoCaptured")

    OnCreate {
      controller = SonyCameraController()
      controller.addListener(self)
      SonyCameraRegistry.controller = controller
      controller.start()
    }

    OnAppEntersForeground {
      controller.refresh()
    }

    OnDestroy {
      controller.removeListener(self)
      controller.close()
      if SonyCameraRegistry.controller === controller {
        SonyCameraRegistry.controller = nil
      }
    }

    Function("getState") {
      controller.statePayload()
    }

    // The options argument is declared in TypeScript and must be accepted here so the
    // documented signature matches runtime. Android supports transport/protocol preference
    // hints and iOS enforces supported wired behavior.
    AsyncFunction("connect") { (options: [String: Any]?) in
      try controller.connectBlocking(options: options ?? [:])
    }

    AsyncFunction("disconnect") {
      controller.disconnectBlocking()
    }

    Function("getDiagnostics") {
      controller.diagnosticsSnapshot()
    }

    Function("clearDiagnostics") {
      controller.clearDiagnostics()
    }

    AsyncFunction("startLiveView") {
      try controller.startLiveView()
      return controller.statePayload()
    }

    AsyncFunction("stopLiveView") {
      controller.stopLiveView()
      return controller.statePayload()
    }

    AsyncFunction("capturePhoto") {
      try controller.capturePhotoBlocking()
    }

    AsyncFunction("capturePreviewFrame") {
      try controller.capturePreviewFrameBlocking()
    }

    View(SonyCameraView.self) {
      Prop("active") { (view: SonyCameraView, active: Bool) in
        view.setActive(active)
      }
    }
  }

  func sonyCameraController(_ controller: SonyCameraController, didChangeState payload: [String: Any]) {
    let state = payload["state"] as? String ?? "unknown"
    let message = payload["message"] as? String ?? ""
    NSLog("[SonyCamera][state] %@: %@", state, message)
    sendEvent("onStateChanged", payload)
  }

  func sonyCameraController(_ controller: SonyCameraController, didAttach payload: [String: Any]) {
    let state = payload["state"] as? String ?? "unknown"
    let message = payload["message"] as? String ?? ""
    NSLog("[SonyCamera][attach] %@: %@", state, message)
    sendEvent("onDeviceAttached", payload)
  }

  func sonyCameraController(_ controller: SonyCameraController, didCapture payload: [String: Any]) {
    NSLog("[SonyCamera][capture] completed")
    sendEvent("onPhotoCaptured", payload)
  }

  func sonyCameraController(_ controller: SonyCameraController, didReceiveFrame jpeg: Data) {}
}

enum SonyCameraRegistry {
  static weak var controller: SonyCameraController?
}
